clustered/server/src/main/java/org/ehcache/clustered/server/state/MessageTracker.java

```diff
@@ -79,17 +83,26 @@ public boolean isEmpty() {
     return nonContiguousMsgIds.isEmpty();
   }
 
+
+  /**
+   * Notify Message tracker that the sync is completed.
+   */
+  public void notifySyncCompleted() {
+    this.isSyncCompleted = true;
+  }
+
   /**
    * Remove the contiguous seen msgIds from the nonContiguousMsgIds and update highestContiguousMsgId
    */
   private void reconcile() {
 
     // This happens when a passive is started after Active has moved on and
     // passive starts to see msgIDs starting from a number > 0.
-    if (highestContiguousMsgId == -1L && nonContiguousMsgIds.size() > 100) {
-      Long min = nonContiguousMsgIds.first();
+    if (highestContiguousMsgId == -1L && nonContiguousMsgIds.size() > 100 && isSyncCompleted) {
+      Long min = nonContiguousMsgIds.last();
       LOGGER.info("Setting highestContiguousMsgId to {} from -1", min);
       highestContiguousMsgId = min;
+      nonContiguousMsgIds.clear();
     }
 
     for (long msgId : nonContiguousMsgIds) {
```

Comment:
I would suggest putting a warning message here in case the set size grows beyond 200 and never comes down.